
# INSinPAL: Large Insertion Detection in Palindromic Fragile Sites

## Description:

INSinPAL is a Snakemake workflow designed to detect insertion events, classified as structural variants (SVs) with a minimum size of 50 base pairs, from mapped paired-end sequencing reads. Given the high number of insertion calls generated by SV callers, INSinPAL focuses specifically on palindromic fragile sites, which are known to undergo genomic rearrangment. This targeted approach enhance the detection of large insertions relavant for rare disease diagnostics (i.e see [*DNA inverted repeats and human disease*](https://www.imrpress.com/journal/FBL/3/4/10.2741/A284)).

INSinPAL considers three type of palindromic fragile sites on one DNA strand:
- Perfect palindromes: Those who exhibit a perfect reverse complemantary sequence matching between both arms.
- Semi-palindromes: Those who have missmatches between both arms, but which achieve a certain matching threshold (85%).
- Inverted-repeats: Those who have spacer betweem both arms of a certain length (maxSize=20) and a matching rate of at least (85%).
Then extract their characteristics and compute the putative fragility of the site to filter them:
- Chromosome name
- Start position (0 based)
- Stop position
- Pal type
- Pal size (both arms + spacer)
- Spacer size
- Mismatch rate (%)
- dAT content (%)
- Recombinogenecity score (from paper below)
- Size group


(Criteria for palindromic sites fragility were selected from: [*long inverted repeats in eukaryotic genomes: Recombinogenic motifs determine genomic plasticity*](https://www.sciencedirect.com/science/article/pii/S0014579306000986))

### Workflow steps

INSinPAL Snakemale workflow will proceed as following:
 - Mine palindromic sites from a reference genome and select fragile ones.
 - Call insertion SVs on sample BAM and merge callers output.
 - Select insertion SVs mapping to palindromic fragile sites.
 - Format VCFs to BEDs and annotate for MEIs, putative inserted sequences source and size.
 - An additional step for BAM formatting exists if mate score tags are absent, needed for INSurVeyo caller.

 



