
# INSinPAL: Large Insertion Detection in Palindromic Fragile Sites

## Description:

INSinPAL is a Snakemake workflow designed to detect insertion events, classified as structural variants (SVs) with a minimum size of 50 base pairs, from mapped paired-end sequencing reads. Given the high number of insertion calls generated by SV callers, INSinPAL focuses specifically on palindromic fragile sites, which are known to undergo genomic rearrangment. This targeted approach enhance the detection of large insertions relavant for rare disease diagnostics (i.e see [*DNA inverted repeats and human disease*](https://www.imrpress.com/journal/FBL/3/4/10.2741/A284)).

Criteria for palindromic fragile sites were selected from: [*long inverted repeats in eukaryotic genomes: Recombinogenic motifs determine genomic plasticity*](https://www.sciencedirect.com/science/article/pii/S0014579306000986).

### Workflow steps

INSinPAL Snakemake workflow will proceed as following:
 - Mine palindromic sites from a reference genome and select fragile ones.
 - Call insertion SVs on sample BAM and merge callers output.
 - Select insertion SVs mapping to palindromic fragile sites.
 - Format VCFs to BEDs and annotate for MEIs, putative inserted sequences source and size.
 - An additional step for BAM formatting exists if mate score tags are absent, needed for INSurVeyo caller.

### Documentation

See ![DAG of jobs](./doc/dag.pdf) for workflow rules structure and/or read the [user guide](./doc/user_guide.md) detailed information about the workflow as setup steps, biological concepts, etc...
 



